[#60144] primerize wp types settings form

Kharonus opened the pull request:
# Ticket
[OP#60144](https://community.openproject.org/work_packages/60144)

# What are you trying to accomplish?
- harmonize look and feel of work package type forms

# What approach did you choose and why?
- remove old form for new and edit
- add primer form component

Kharonus committed:
[#60144] primerize wp types settings form

- https://community.openproject.org/work_packages/60144
- remove old form for new and edit
- add primer form component

Kharonus committed:
[#60144] fixed types controller test

NobodysNightmare commented on app/forms/work_packages/types/settings_form.rb:
```diff
@@ -0,0 +1,108 @@
+module WorkPackages
+  module Types
+    class SettingsForm < ApplicationForm
+      form do |subject_form|
```
:yellow_circle: Probably a copy & paste leftover, but this is the `settings_form`, not the `subject_form`.

Kharonus replied:
❗ nooo, how can this even happen ❗

NobodysNightmare reviewed (approved):
Changes look (literally) nice.

(Obviously assuming the failing spec will be fixed)

Kharonus committed:
[#60144] retain and submit copy workflow value

- fetch correct parameter for copy workflow value on form submission
- initialize form with this parameter if available
- fix feature spec with new UI

NobodysNightmare commented on app/views/types/new.html.erb:
```diff
@@ -39,8 +39,10 @@ See COPYRIGHT and LICENSE files for more details.
+  render WorkPackages::Types::SettingsComponent.new(
+    @type,
+    copy_workflow_from: params.dig(:type, :copy_workflow_from),
+    params:
```
:red_circle: Passing `params` into the component is a debugging left-over, isn't it?

Kharonus replied:
yes, definitely, good catch!

NobodysNightmare commented on app/components/work_packages/types/settings_component.rb:
```diff
@@ -0,0 +1,52 @@
+      include OpTurbo::Streamable
+
+      def initialize(model, **options)
+        @copy_workflow_from = options[:copy_workflow_from].to_i
```
:green_circle: Maybe it's just me, but this `to_i` feels wrong to me. In the end it's just a matter of passing the value through from the outside, but maybe I'd prefer if the conversion took place outside of `SettingsComponent` (in the place that fills this from `params`, where we know that things are only _stringly_ typed.

Also as an aside: We _could_ make use of named arguments in this method. Something along the lines of:

```ruby
def initialize(model, copy_workflow_from: nil, **opts)
  # <--- use copy_workflow_from here
  super(model, **opts) # bonus: The superclass constructor will not receive the named argument that we consumed
end
```

mereghost replied:
I'd add that I'd prefer to use `Integer` but that will raise an error rather than failing gracefully.

I'm with @NobodysNightmare here. I'd prefer to have the input sanitized beforehand.

Kharonus replied:
Maybe there is something wrong a bit deeper. When starting to develop this functionality I started using strings for the type id. from params we also get strings, other records use string as their id. but for some reason, the `type.id` is an integer.

so, I have two options
- sanitizing this when first reading from params
- have the comparison function work to safely compare strings with integer ids

WDYT?

NobodysNightmare replied:
I'd prefer to sanitize when reading from `params`. The latter is what a Rails method would probably try to do (e.g. `Type.find(...)`. But if I can choose, I try to write code that uses the correct types early.

mereghost commented on app/components/work_packages/types/settings_component.rb:
```diff
@@ -0,0 +1,52 @@
+        super
+      end
+
+      def form_options
```
🟢 I suspect that I probably was the first to make this option hashes with ternaries, but I think that having them as separate methods will make the intent a bit clearer.

```ruby
def form_options
  return create_form_options if model.new_record?
  
  update_form_options
end

private

attr_reader :copy_workflow_from

def update_form_options = { url: type_path(id: model.id), method: :patch, model:, copy_workflow_from: }
```

mereghost commented on app/controllers/types_controller.rb:
```diff
@@ -62,19 +62,17 @@ def edit
     CreateTypeService
       .new(current_user)
-      .call(permitted_type_params, copy_workflow_from: params[:copy_workflow_from]) do |call|
+      .call(permitted_type_params, copy_workflow_from: params.dig(:type, :copy_workflow_from)) do |call|
```
🟡 Am I missing something here (most probably yes) but since we are feeding the all the permitted type params, why do we need to the `copy_workflow_from` as a special ❄️?

Kharonus replied:
`copy_workflow_from` is no property from the type record.

previously (as you can see from the change) it was passed not even inside `params|type` but directly on top level params.

Usually, having something like this, I'd rework the whole form to use a form model, as this is apparently what the create service wants. But for sake of simplicity I tried to keep the previous approach alive. But with the primer forms I no longer was able to define the form input field outside the "model".

mereghost commented on app/controllers/types_controller.rb:
```diff
@@ -62,19 +62,17 @@ def edit
       @type = call.result
 
       call.on_success do
         redirect_to_type_tab_path(@type, t(:notice_successful_create))
```
🟡 If you are dealing with Turbo Frames, it is recommended to redirect with a `:see_other` (303) status

mereghost reviewed (approved):
Left a couple of questions but nothing that should block the merge. :shipit:

Kharonus committed:
[#60144] sanitizing parameter input into settings component

- use maybe to convert from parameter value to integer
- split create and update form options
- add 303 status to redirection in types controller
